builtin/rebase.c

```diff
@@ -165,6 +196,86 @@ int cmd_rebase(int argc, const char **argv, const char *prefix)
 		options.state_dir = merge_dir();
 		break;
 	}
+	if (!options.root) {
+		if(!argc)
+			die("TODO: handle @{upstream}");
+		else {
+			options.upstream_name = argv[1];
+			if (!strcmp(options.upstream_name, "-"))
+				options.upstream_name = "@{-1}";
+		}
+		options.upstream = peel_committish(options.upstream_name);
+		if (!options.upstream)
+			die(_("invalid upstream '%s'"), options.upstream_name);
+	} else
+		die("TODO: upstream for --root");
 
-	return !!run_specific_rebase(&options);
+	/* Make sure the branch to rebase onto is valid. */
+	options.onto_name = (const char *) options.onto ?
+		(const char *) options.onto : options.upstream_name;
+	if (strstr(options.onto_name, "...")) {
+		die("TODO");
+	} else {
+		options.onto = peel_committish(options.onto_name);
+		if (!options.onto)
+			die(_("Does not point to a valid commit '%s'"),
+				options.onto_name);
+	}
+
+	/*
+ 	* If the branch to rebase is given, that is the branch we will rebase
+ 	* branch_name -- branch/commit being rebased, or HEAD (already detached)
+ 	* orig_head -- commit object name of tip of the branch before rebasing
+ 	* head_name -- refs/heads/<that-branch> or "detached HEAD"
+ 	*/
+ 	if (argc > 1)
+		 die ("TODO: handle switch_to");
+	else {
+		/* Do not need to switch branches, we are already on it. */
+		options.head_name =
+			xstrdup_or_null(resolve_ref_unsafe("HEAD", 0, NULL,
+					 &flags));
+		if (!options.head_name)
+			die(_("No such ref: %s"), "HEAD");
+		if (flags & REF_ISSYMREF) {
+			if (!skip_prefix(options.head_name,
+					 "refs/heads/", &branch_name))
+				branch_name = options.head_name;
+
+		} else {
+			options.head_name = xstrdup("detached HEAD");
+			branch_name = "HEAD";
+		}
+		if (get_oid("HEAD", &options.orig_head))
+			die(_("Could not resolve HEAD to a revision"));
+	}
+
+	/* Detach HEAD and reset the tree */
+	if (!quiet)
+		printf("First, rewinding head to replay your work on top of it...");
+
+	strbuf_addf(&msg, "rebase: checkout %s", options.onto_name);
+	if (detach_head_to(&options.onto->object.oid, "checkout", msg.buf))
+		die(_("Could not detach HEAD"));
+	strbuf_release(&msg);
+	if (update_ref("rebase", "ORIG_HEAD", &options.orig_head, NULL, 0,
+	 	       UPDATE_REFS_MSG_ON_ERR) < 0)
+		die(_("Could not update ORIG_HEAD to '%s'"),
+		    oid_to_hex(&options.orig_head));
+
+	if (!options.root)
+		strbuf_addf(&revisions, "%s..%s",
+			    oid_to_hex(&options.onto->object.oid),
+			    oid_to_hex(&options.orig_head));
+	else if (restrict_revision)
+		strbuf_addf(&revisions, "%s..%s",
+			    oid_to_hex(&options.onto->object.oid),
+			    oid_to_hex(&options.orig_head));
```

Comment:
This is still missing the default case:

```c
else
    strbuf_addf(&revisions, "%s..%s",
        oid_to_hex(&options.upstream->object.oid),
        oid_to_hex(&options.orig_head));
```

BTW a more concise way to write this would use the `<condition> ? <then> : <else>` notation, i.e.

```c
strbuf_addf(&revisions, "%s..%s",
    !options.root ? oid_to_hex(&options.onto->object.oid) :
        (restrict_revision ? restrict_revision : oid_to_hex(&options.upstream->object.oid)),
    oid_to_hex(&options.orig_head));
```

This has the advantage of being less repetitive (i.e. it heeds the [DRY principle](https://en.wikipedia.org/wiki/Don't_repeat_yourself)), but it is arguably harder to read for developers who are not so familiar with that C construct (although I have to admit that it is present in Javascript, too).

Reply:
Oh, and we definitely need to assign the result to the rebase_options field:

```c
options.revisions = revisions.buf
```

and then use it in `run_specific_rebase()` to set that environment variable in the `env` array.